src/Adapter/Order/CommandHandler/DeleteCartRuleFromOrderHandler.php

```diff
@@ -85,12 +100,74 @@ public function handle(DeleteCartRuleFromOrderCommand $command)
 
         try {
             // Delete Order Cart Rule and update Order
+            $cartRule = new CartRule($orderCartRule->id_cart_rule);
             $orderCartRule->softDelete();
             $cart->removeCartRule($orderCartRule->id_cart_rule);
 
-            $this->orderAmountUpdater->update($order, $cart, $orderCartRule->id_order_invoice);
+            // If cart rule was a gift product we must update an OrderDetail manually
+            $giftOrderDetail = $this->getGiftOrderDetail($order, $cartRule);
+            if (null !== $giftOrderDetail) {
+                $newQuantity = ((int) $giftOrderDetail->product_quantity) - 1;
+
+                /*
+                 * Note: we are lucky the stock updates happens smoothly, it was not re-injected by removing the CartRule
+                 * because Cart doesn't update the stock but it will while we update the OrderDetail Lucky
+                 * for us it's the same difference of 1 so it's all accurate in the end
+                 */
+                // This calls the OrderAmountUpdater internally so no need to perform both calls
+                $this->orderProductQuantityUpdater->update(
+                    $order,
+                    $giftOrderDetail,
+                    $newQuantity,
+                    $giftOrderDetail->id_order_invoice != 0 ? new OrderInvoice($giftOrderDetail->id_order_invoice) : null,
```

Comment:
Unless `id_order_invoice` can be a string
```suggestion
                    $giftOrderDetail->id_order_invoice !== 0 ? new OrderInvoice($giftOrderDetail->id_order_invoice) : null,
```
Or else
```suggestion
                    $giftOrderDetail->id_order_invoice == 0 ? null : new OrderInvoice((int) $giftOrderDetail->id_order_invoice),
```

Reply:
Yes in legacy objects most of the time int are string I added casting and strict equality